sklearn/metrics/cluster/_unsupervised.py

```diff
@@ -131,15 +132,25 @@ def _silhouette_reduce(D_chunk, start, labels, label_freqs):
     label_freqs : array-like
         Distribution of cluster labels in ``labels``.
     """
+    n_chunk_samples = D_chunk.shape[0]
     # accumulate distances from each sample to each cluster
-    clust_dists = np.zeros((len(D_chunk), len(label_freqs)), dtype=D_chunk.dtype)
-    for i in range(len(D_chunk)):
+    clust_dists = np.zeros((n_chunk_samples, len(label_freqs)), dtype=D_chunk.dtype)
+    D_chunk_is_sparse = issparse(D_chunk)
+    for i in range(n_chunk_samples):
+        if D_chunk_is_sparse:
+            current_chunk = D_chunk.getrow(i)
```

Comment:
It is unfortunate that `getrow` always returns a copy, even for CSR matrices: https://docs.scipy.org/doc//scipy/reference/generated/scipy.sparse.csr_array.getrow.html?highlight=getrow#scipy.sparse.csr_array.getrow

For me, I am okay with optimizing for `csr`, because most of scikit-learn is optimized for CSR matrices:

```python
    if D_chunk_is_sparse and D_chunk.format != "csr":
        D_chunk = D_chunk.tocsr()

    for i in range(n_chunk_samples):
        if D_chunk_is_sparse:
            indptr = D_chunk.indptr
            indices = D_chunk.indices[indptr[i] : indptr[i + 1]]
            sample_weights = D_chunk.data[indptr[i] : indptr[i + 1]]
            sample_labels = np.take(labels, indices)
```

If the precomputed matrix is CSR, `tocsr` is a no-op and getting the `sample_labels` and `sample_weights` do not make a copy anymore.

Reply:
Okay. I will make the necessary changes, so that we optimize for CSR matrices.